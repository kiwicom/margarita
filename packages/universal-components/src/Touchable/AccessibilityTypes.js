// @flow

type AccessibilityTraitsValues =
  | 'none'
  | 'button'
  | 'link'
  | 'header'
  | 'search'
  | 'image'
  | 'selected'
  | 'plays'
  | 'key'
  | 'text'
  | 'summary'
  | 'disabled'
  | 'frequentUpdates'
  | 'startsMedia'
  | 'adjustable'
  | 'allowsDirectInteraction'
  | 'pageTurn';

// See: https://facebook.github.io/react-native/docs/accessibility.html
export type AccessibilityProps = {|
  accessible?: boolean,
  accessibilityLabel?: string,
  accessibilityTraits?:  // iOS only
    | AccessibilityTraitsValues
    | $ReadOnlyArray<AccessibilityTraitsValues>,
  accessibilityViewIsModal?: boolean, // iOS only
  onAccessibilityTap?: () => void, // iOS only
  onMagicTap?: () => void, // iOS only
  accessibilityComponentType?:  // Android only
    | 'none'
    | 'button'
    | 'radiobutton_checked'
    | 'radiobutton_unchecked',
  accessibilityLiveRegion?: 'none' | 'polite' | 'assertive', // Android only
  importantForAccessibility?: 'auto' | 'yes' | 'no' | 'no-hide-descendants', // Android only
|};
