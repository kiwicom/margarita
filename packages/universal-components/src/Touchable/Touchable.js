// @flow

import * as React from 'react';
import {
  Platform,
  TouchableNativeFeedback,
  TouchableOpacity,
  View,
} from 'react-native';

import type { AccessibilityProps as AccessibilityPropsType } from './AccessibilityTypes';
import type { StylePropType } from '../PlatformStyleSheet/StyleTypes';

type Props = {|
  +children: React.Node,
  +onPress: () => void,
  +disabled: boolean,
  +width?: number,
  +style?: StylePropType,
  +onLongPress?: () => void,
  +delayPressIn?: number,
  // This will disable ripple effect completely and it will fallback to the
  // opacity behavior.
  +noRipple?: boolean,
  // Should the ripple render outside of the view bounds?
  +borderlessRipple?: boolean,
  +rippleColor?: string,
  +testID?: string,
  ...AccessibilityPropsType,
|};

/**
 * Touchable renders a touchable that looks native on both iOS and Android.
 * It provides an abstraction on top of TouchableNativeFeedback and
 * TouchableOpacity. On iOS you can pass the props of TouchableOpacity, on
 * Android pass the props of TouchableNativeFeedback.
 */
export default class Touchable extends React.Component<Props> {
  static defaultProps = {
    borderlessRipple: false,
    rippleColor: 'rgba(0, 0, 0, 0.32)',
    disabled: false,
    accessibilityTraits: 'button',
    accessibilityComponentType: 'button',
  };

  /**
   * TouchableNativeFeedback.Ripple causes a crash on old Android versions,
   * therefore only enable it on Android Lollipop and above.
   */
  supportsRippleEffect = () => {
    const { noRipple } = this.props;
    if (noRipple === true) {
      return false;
    }
    return Platform.OS === 'android' && Platform.Version >= 21;
  };

  render() {
    const {
      children: childrenProps,
      onPress,
      style,
      onLongPress,
      delayPressIn,
      noRipple,
      borderlessRipple,
      rippleColor,
      testID,
      width,
      disabled,
      ...accessiblityProps
    } = this.props;

    const touchableNativeFeedbackProps = {
      onPress,
      onLongPress,
      delayPressIn,
      noRipple,
      testID,
      ...accessiblityProps,
    };

    const touchableOpacityProps = {
      disabled,
      onPress,
      onLongPress,
      delayPressIn,
      testID,
    };

    // React.Children.only is only necessary for TouchableNativeFeedback
    // but since we are building for both platforms it is here as well
    // to discover this mistake as soon as possible
    const children = React.Children.only(childrenProps);

    // All touchables on Android should have the ripple effect according to
    // platform design guidelines.
    if (this.supportsRippleEffect()) {
      let useForeground = TouchableNativeFeedback.canUseNativeForeground();
      if (borderlessRipple) {
        useForeground = false;
      }

      return (
        <TouchableNativeFeedback
          {...touchableNativeFeedbackProps}
          style={{ width }}
          useForeground={useForeground}
          background={TouchableNativeFeedback.Ripple(
            rippleColor,
            borderlessRipple,
          )}
        >
          <View style={style}>{children}</View>
        </TouchableNativeFeedback>
      );
    }

    return (
      <TouchableOpacity
        activeOpacity={0.5}
        style={{ width }}
        {...touchableOpacityProps}
      >
        <View style={style}>{children}</View>
      </TouchableOpacity>
    );
  }
}
