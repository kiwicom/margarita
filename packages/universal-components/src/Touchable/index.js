// @flow

export { default as Touchable } from './Touchable';
