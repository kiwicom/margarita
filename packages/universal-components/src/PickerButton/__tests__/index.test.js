// @flow

import * as React from 'react';
import { render, fireEvent } from 'react-native-testing-library';
import snapshotDiff from 'snapshot-diff';

import { PickerButton } from '..';

import { Touchable } from '../../Touchable';

const noop = jest.fn();

describe('PickerButton', () => {
  it('should contain placeholder text', () => {
    const placeholder = 'Placeholder';
    const { getByText } = render(<PickerButton placeholder={placeholder} />);
    expect(getByText(placeholder)).toBeDefined();
  });

  it('should contain value text', () => {
    const value = 'Value';
    const { getByText } = render(<PickerButton value={value} />);
    expect(getByText(value)).toBeDefined();
  });

  it('should execute onPress function', () => {
    const onPress = jest.fn();
    const { getByType } = render(<PickerButton onPress={onPress} />);
    fireEvent.press(getByType(Touchable));
    expect(onPress).toHaveBeenCalledTimes(1);
  });

  it('should match snapshot diff', () => {
    const base = <PickerButton />;
    const extend = (
      <PickerButton
        value="Value"
        onPress={noop}
        placeholder="Select"
        iconName="chevron-down"
      />
    );
    expect(snapshotDiff(base, extend)).toMatchSnapshot();
  });
});
