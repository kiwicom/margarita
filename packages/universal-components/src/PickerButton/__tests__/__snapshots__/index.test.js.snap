// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PickerButton should match snapshot diff 1`] = `
"Snapshot Diff:
- <PickerButton />
+ <PickerButton iconName=\\"chevron-down\\" onPress={[Function mockConstructor]} placeholder=\\"Select\\" value=\\"Value\\" />

@@ -1,8 +1,33 @@
  <View
+   accessible={true}
+   isTVSelectable={true}
+   onResponderGrant={[Function bound touchableHandleResponderGrant]}
+   onResponderMove={[Function bound touchableHandleResponderMove]}
+   onResponderRelease={[Function bound touchableHandleResponderRelease]}
+   onResponderTerminate={[Function bound touchableHandleResponderTerminate]}
+   onResponderTerminationRequest={[Function bound touchableHandleResponderTerminationRequest]}
+   onStartShouldSetResponder={[Function bound touchableHandleStartShouldSetResponder]}
+   style={
+     Object {
+       \\"opacity\\": 1,
+     }
+   }
+ >
+   <View
      style={
+       Array [
+         undefined,
          Object {
+           \\"width\\": undefined,
+         },
+       ]
+     }
+   >
+     <View
+       style={
+         Object {
            \\"alignItems\\": \\"center\\",
            \\"backgroundColor\\": \\"#e8edf1\\",
            \\"borderRadius\\": 6,
            \\"flex\\": 1,
            \\"flexBasis\\": 44,
@@ -24,18 +49,18 @@
              },
              Array [
                Object {
                  \\"color\\": \\"#171b1e\\",
                  \\"paddingEnd\\": 12,
-           },
-           Object {
-             \\"color\\": \\"#bac7d5\\",
                },
+               false,
              ],
            ]
          }
-   />
+       >
+         Value
+       </Text>
        <Text
          style={
            Array [
              Object {
                \\"fontFamily\\": \\"orbit-icons\\",
@@ -51,8 +76,10 @@
              },
              undefined,
            ]
          }
        >
-     
+         
        </Text>
+     </View>
+   </View>
  </View>"
`;
